### DISCUSSION TITLE
Environment locking, bug fixes, stability improvements + various other changes.

### DISCUSSION THREAD

**Discussion by peterhessey** [microsoft/InnerEye-DeepLearning#742] [General]
## What's Changed
* Correct path for ensemble models by @JonathanTripp in https://github.com/microsoft/InnerEye-DeepLearning/pull/472
* Fix for SDK regression bug, minor doc updates by @ant0nsc in https://github.com/microsoft/InnerEye-DeepLearning/pull/475
* Set l_rate correctly and ensure no overlap between subjects in dataset splits in CovidHierarchicalModel config by @Shruthi42 in https://github.com/microsoft/InnerEye-DeepLearning/pull/467
* Update documentation for submodules by @ant0nsc in https://github.com/microsoft/InnerEye-DeepLearning/pull/481
* Multi-node checkpoint recovery fix by @melanibe in https://github.com/microsoft/InnerEye-DeepLearning/pull/478
* Check bool parameter is either true or false (after lower) by @JonathanTripp in https://github.com/microsoft/InnerEye-DeepLearning/pull/482
* Better handling of missing seriesId by @JonathanTripp in https://github.com/microsoft/InnerEye-DeepLearning/pull/488
* Minor bug fixes in CovidHierarchicalModel by @Shruthi42 in https://github.com/microsoft/InnerEye-DeepLearning/pull/490
* Better handling of outliers by @JonathanTripp in https://github.com/microsoft/InnerEye-DeepLearning/pull/489
* Fix for stuck test set inference for LightningContainer models by @ant0nsc in https://github.com/microsoft/InnerEye-DeepLearning/pull/494
* Regression test coverage for AzureML runs by @ant0nsc in https://github.com/microsoft/InnerEye-DeepLearning/pull/492
* Bug fix for regression test by @ant0nsc in https://github.com/microsoft/InnerEye-DeepLearning/pull/496
* Remove 2 files that cause PR build to fail non-reproducibly by @ant0nsc in https://github.com/microsoft/InnerEye-DeepLearning/pull/500
* Fix timeouts when downloading multiple checkpoint files by @ant0nsc in https://github.com/microsoft/InnerEye-DeepLearning/pull/498
* Reduce AML snapshot size by skipping test folders by @ant0nsc in https://github.com/microsoft/InnerEye-DeepLearning/pull/497
* Use best epoch for model comparison by @JonathanTripp in https://github.com/microsoft/InnerEye-DeepLearning/pull/495
* Fix bug in all_azure_dataset_ids and all_dataset_mountpoints by @Shruthi42 in https://github.com/microsoft/InnerEye-DeepLearning/pull/466
* Allow cross validation with 'bring your own' Lightning models (without ensemble building) by @dumbledad in https://github.com/microsoft/InnerEye-DeepLearning/pull/483
* Comment out glaucoma job by @JonathanTripp in https://github.com/microsoft/InnerEye-DeepLearning/pull/520
* Split validation and test infer config by @JonathanTripp in https://github.com/microsoft/InnerEye-DeepLearning/pull/502
* Added ability to run segmentation inference module in the test data without or partial ground truth files.   by @asantamariapang in https://github.com/microsoft/InnerEye-DeepLearning/pull/465
* Partial ground truth inference PR (465) missed two things by @dumbledad in https://github.com/microsoft/InnerEye-DeepLearning/pull/522
* Enable store_dataset_sample by @javier-alvarez in https://github.com/microsoft/InnerEye-DeepLearning/pull/525
* Better Ensemble Child Inference Defaults by @JonathanTripp in https://github.com/microsoft/InnerEye-DeepLearning/pull/533
* Update PL to 1.3.8 by @melanibe in https://github.com/microsoft/InnerEye-DeepLearning/pull/531
* Update README.md by @ktakeda1 in https://github.com/microsoft/InnerEye-DeepLearning/pull/511
* Fixes for mounting and matplotlib problems by @ant0nsc in https://github.com/microsoft/InnerEye-DeepLearning/pull/515
* Dropping windows tests, but keeping cred-scan by @dumbledad in https://github.com/microsoft/InnerEye-DeepLearning/pull/542
* Run inference using checkpoints from registered models by @Shruthi42 in https://github.com/microsoft/InnerEye-DeepLearning/pull/509
* Enable a disabled test by @ant0nsc in https://github.com/microsoft/InnerEye-DeepLearning/pull/536
* Print warning if inference is disabled but comparison requested by @JonathanTripp in https://github.com/microsoft/InnerEye-DeepLearning/pull/537
* Update to fix issues in daily build by @ant0nsc in https://github.com/microsoft/InnerEye-DeepLearning/pull/545
* Update Covid configs by @Shruthi42 in https://github.com/microsoft/InnerEye-DeepLearning/pull/526
* Environment and hello_world_model documentation updated by @vale-salvatelli in https://github.com/microsoft/InnerEye-DeepLearning/pull/546
* Fix: pl_find_unused_parameters was no longer used by @ant0nsc in https://github.com/microsoft/InnerEye-DeepLearning/pull/547
* Minor changes to CovidModel config parameters and updated report by @Shruthi42 in https://github.com/microsoft/InnerEye-DeepLearning/pull/554
* Fix incomplete test data module setup in Lightning inference by @dccastro in https://github.com/microsoft/InnerEye-DeepLearning/pull/553
* Moving InnerEye's Azure code to hi-ml package by @ant0nsc in https://github.com/microsoft/InnerEye-DeepLearning/pull/548
* Fix learning rate parameter in SSLContainer by @Shruthi42 in https://github.com/microsoft/InnerEye-DeepLearning/pull/557
* Use SSLEncoder when building model in CovidModel config by @Shruthi42 in https://github.com/microsoft/InnerEye-DeepLearning/pull/558
* Generalize SSL functionality to work on other datasets by @vale-salvatelli in https://github.com/microsoft/InnerEye-DeepLearning/pull/555
* Adding Active Label Cleaning code by @melanibe in https://github.com/microsoft/InnerEye-DeepLearning/pull/559
* Update environment.yml to `hi-ml-azure>=0.1` by @dccastro in https://github.com/microsoft/InnerEye-DeepLearning/pull/566
* Updating pillow to fix component governance issue by @vale-salvatelli in https://github.com/microsoft/InnerEye-DeepLearning/pull/567
* Adding pre-commit hooks and black formatting by @vale-salvatelli in https://github.com/microsoft/InnerEye-DeepLearning/pull/560
* Adjust to new namespaces by @ant0nsc in https://github.com/microsoft/InnerEye-DeepLearning/pull/572
* Document Segmentation Model Evaluation by @JonathanTripp in https://github.com/microsoft/InnerEye-DeepLearning/pull/544
* Remove the stdout.txt file by @ant0nsc in https://github.com/microsoft/InnerEye-DeepLearning/pull/576
* Switching batch time loading diagnostics to hi-ml by @ant0nsc in https://github.com/microsoft/InnerEye-DeepLearning/pull/577
* Update hello_world_model.md by @maxilse in https://github.com/microsoft/InnerEye-DeepLearning/pull/580
* Fix for AzureML environment problem by @ant0nsc in https://github.com/microsoft/InnerEye-DeepLearning/pull/587
* Enable overriding AzureConfig parameters from a LightningContainer by @dccastro in https://github.com/microsoft/InnerEye-DeepLearning/pull/589
* Fix for invalid path_on_compute problem by @ant0nsc in https://github.com/microsoft/InnerEye-DeepLearning/pull/593
* Fixing SSL recovery, attempt 2 by @ant0nsc in https://github.com/microsoft/InnerEye-DeepLearning/pull/584
* Use PIL.PngImagePlugin to load pngs by @JonathanTripp in https://github.com/microsoft/InnerEye-DeepLearning/pull/588
* Make pytorch run non-deterministically by default, upgrade to AML SDK 1.36 by @ant0nsc in https://github.com/microsoft/InnerEye-DeepLearning/pull/594
* Add cudatoolkit=11.1 to environment.yml by @dccastro in https://github.com/microsoft/InnerEye-DeepLearning/pull/596
* Add histopathology module and add hi-ml as submodule  by @mebristo in https://github.com/microsoft/InnerEye-DeepLearning/pull/603
* Add `--pl_deterministic` to build training jobs by @dccastro in https://github.com/microsoft/InnerEye-DeepLearning/pull/605
* Bug fix: deployed models and training code use different versions of hi-ml by @ant0nsc in https://github.com/microsoft/InnerEye-DeepLearning/pull/606
* Bug fix: When using local folders, datasets are downloaded nevertheless by @ant0nsc in https://github.com/microsoft/InnerEye-DeepLearning/pull/604
* Mebristo/move histopathology update by @mebristo in https://github.com/microsoft/InnerEye-DeepLearning/pull/613
* Downloading checkpoints from AML if not found on disk by @ant0nsc in https://github.com/microsoft/InnerEye-DeepLearning/pull/614
* Enabling distributed training for SSL online evaluator by @ant0nsc in https://github.com/microsoft/InnerEye-DeepLearning/pull/612
* Add more histopathology configs by @mebristo in https://github.com/microsoft/InnerEye-DeepLearning/pull/616
* Small cleanups for checkpoint loading by @ant0nsc in https://github.com/microsoft/InnerEye-DeepLearning/pull/615
* Simplify setting callbacks in LightningContainers by @ant0nsc in https://github.com/microsoft/InnerEye-DeepLearning/pull/617
* Adding pipeline cache for conda environment by @ant0nsc in https://github.com/microsoft/InnerEye-DeepLearning/pull/618
* Vsalva/deepmil panda by @vale-salvatelli in https://github.com/microsoft/InnerEye-DeepLearning/pull/619
* Upgrade to Pytorch Lightning 1.5.5 by @ant0nsc in https://github.com/microsoft/InnerEye-DeepLearning/pull/591
* Bug fix: multi-GPU jobs on a VM use wrong folders by @ant0nsc in https://github.com/microsoft/InnerEye-DeepLearning/pull/622
* Enable tiling non-PANDA WSI datasets by @dccastro in https://github.com/microsoft/InnerEye-DeepLearning/pull/621
* Bug fix: SSL on multiple nodes used wrong LR scheduler by @ant0nsc in https://github.com/microsoft/InnerEye-DeepLearning/pull/628
* SSL Online evaluator: save checkpoints without DDP wrapper by @ant0nsc in https://github.com/microsoft/InnerEye-DeepLearning/pull/623
* Upgrade to PyTorch 1.10 by @ant0nsc in https://github.com/microsoft/InnerEye-DeepLearning/pull/585
* Moving *.nii.gz from git lfs to git to simplify the HelloWorld test by @ant0nsc in https://github.com/microsoft/InnerEye-DeepLearning/pull/632
* fixing PandaInnereyeSSLMIL by @vale-salvatelli in https://github.com/microsoft/InnerEye-DeepLearning/pull/625
* Improve recovery of preempted jobs by @ant0nsc in https://github.com/microsoft/InnerEye-DeepLearning/pull/633
* Plotting heatmap and thumbnails for test PANDA slides  by @harshita-s in https://github.com/microsoft/InnerEye-DeepLearning/pull/634
* Fix length problem in SSL Cosine LR by @ant0nsc in https://github.com/microsoft/InnerEye-DeepLearning/pull/638
* Panda SSL tile selection and binary evaluation by @harshita-s in https://github.com/microsoft/InnerEye-DeepLearning/pull/635
* SimCLR recovery test by @maxilse in https://github.com/microsoft/InnerEye-DeepLearning/pull/643
* Cancel queued AzureML jobs when starting a PR build by @ant0nsc in https://github.com/microsoft/InnerEye-DeepLearning/pull/640
* Enabling DeepSMILE on large encoded datasets by @vale-salvatelli in https://github.com/microsoft/InnerEye-DeepLearning/pull/637
* Add class-wise metrics logging and confusion matrix to DeepMIL by @harshita-s in https://github.com/microsoft/InnerEye-DeepLearning/pull/647
* Workaround for bug in PL 1.5.5: CombinedLoader cannot be used with DDP for training data by @ant0nsc in https://github.com/microsoft/InnerEye-DeepLearning/pull/646
* Fix in _convert_to_tensor_if_necessary so that also np.array get converted correctly by @maxilse in https://github.com/microsoft/InnerEye-DeepLearning/pull/649
* Fix for stuck Linux build: Move pytest to Windows by @ant0nsc in https://github.com/microsoft/InnerEye-DeepLearning/pull/652
* Re-enable Linux build by @ant0nsc in https://github.com/microsoft/InnerEye-DeepLearning/pull/655
* Add dropout to DeepMIL and fix feature extractor setup by @dccastro in https://github.com/microsoft/InnerEye-DeepLearning/pull/653
* Enable fine-tuning in Deepmil by @harshita-s in https://github.com/microsoft/InnerEye-DeepLearning/pull/650
* Fix DeepMIL for TCGA CRCK dataset by @maxilse in https://github.com/microsoft/InnerEye-DeepLearning/pull/659
* Add subsampling transform and mean pooling by @dccastro in https://github.com/microsoft/InnerEye-DeepLearning/pull/656
* Replace RadIO with TorchIO for patch-based inference by @fepegar in https://github.com/microsoft/InnerEye-DeepLearning/pull/666
* Fix DeepMIL metrics bug by @dccastro in https://github.com/microsoft/InnerEye-DeepLearning/pull/674
* Updated README.md by @ktakeda1 in https://github.com/microsoft/InnerEye-DeepLearning/pull/680
* Add FP and TN to DeepMIL outputs and configure outputs for multi-class classification by @harshita-s in https://github.com/microsoft/InnerEye-DeepLearning/pull/679
* Print out name of log level, not only the number by @fepegar in https://github.com/microsoft/InnerEye-DeepLearning/pull/678
* Fix missing separator error in docs Makefile by @fepegar in https://github.com/microsoft/InnerEye-DeepLearning/pull/683
* Fix Markdown lint errors [skip ci] by @fepegar in https://github.com/microsoft/InnerEye-DeepLearning/pull/684
* Pad model outputs if they are smaller than the inputs by @fepegar in https://github.com/microsoft/InnerEye-DeepLearning/pull/681
* Clean up legacy code by @javier-alvarez in https://github.com/microsoft/InnerEye-DeepLearning/pull/671
* Replace data quality folder with a link to the commit by @javier-alvarez in https://github.com/microsoft/InnerEye-DeepLearning/pull/692
* Ensure the shape of input patches is compatible with model constraints by @fepegar in https://github.com/microsoft/InnerEye-DeepLearning/pull/682
* Update TorchIO version by @fepegar in https://github.com/microsoft/InnerEye-DeepLearning/pull/677
* DOC: Improve setup for running the HelloWorld model in AzureML by @ant0nsc in https://github.com/microsoft/InnerEye-DeepLearning/pull/693
* Specify `formatter_class=argparse.ArgumentDefaultsHelpFormatter` in `generic_parsing.py` to show default values for arguments in help message by @JacopoTeneggi in https://github.com/microsoft/InnerEye-DeepLearning/pull/689
* Fix some Sphinx warnings by @fepegar in https://github.com/microsoft/InnerEye-DeepLearning/pull/699
* Automatically and linearly scale the learning rate of the SSL encoder to the number of GPUS by @maxilse in https://github.com/microsoft/InnerEye-DeepLearning/pull/667
* Import submodules to sys.path to fix autodoc's warning by @JacopoTeneggi in https://github.com/microsoft/InnerEye-DeepLearning/pull/704
* Update hi-ml submodule by @fepegar in https://github.com/microsoft/InnerEye-DeepLearning/pull/709
* Deprecate manual changelog by @fepegar in https://github.com/microsoft/InnerEye-DeepLearning/pull/707
* Ignore Sphinx build folders by @fepegar in https://github.com/microsoft/InnerEye-DeepLearning/pull/698
* ENH: Enforce prefixes for pull request titles by @fepegar in https://github.com/microsoft/InnerEye-DeepLearning/pull/713
* STYLE: Trim trailing whitespaces by @fepegar in https://github.com/microsoft/InnerEye-DeepLearning/pull/715
* STYLE: Ensure line breaks at the end of all files by @fepegar in https://github.com/microsoft/InnerEye-DeepLearning/pull/717
* DOC: Add a smoke-test model for AzureML setup by @ant0nsc in https://github.com/microsoft/InnerEye-DeepLearning/pull/708
* ENH: Customized VSCode Window Color by @ant0nsc in https://github.com/microsoft/InnerEye-DeepLearning/pull/712
* DOC: Add part of the API to the Sphinx docs by @fepegar in https://github.com/microsoft/InnerEye-DeepLearning/pull/718
* BUG: Fix broken build pipeline by updating package versions by @peterhessey in https://github.com/microsoft/InnerEye-DeepLearning/pull/725
* ENH: Add a structured template for issues by @ant0nsc in https://github.com/microsoft/InnerEye-DeepLearning/pull/727
* ENH: Auto add issues to GitHub Project by @peterhessey in https://github.com/microsoft/InnerEye-DeepLearning/pull/721
* ENH: Add Environment Locking by @peterhessey in https://github.com/microsoft/InnerEye-DeepLearning/pull/735
* DOC: Improve documentation for dataset creation by @ant0nsc in https://github.com/microsoft/InnerEye-DeepLearning/pull/723
* STYLE: Fix pre-commit errors by @fepegar in https://github.com/microsoft/InnerEye-DeepLearning/pull/736

## New Contributors
* @ktakeda1 made their first contribution in https://github.com/microsoft/InnerEye-DeepLearning/pull/511
* @vale-salvatelli made their first contribution in https://github.com/microsoft/InnerEye-DeepLearning/pull/546
* @maxilse made their first contribution in https://github.com/microsoft/InnerEye-DeepLearning/pull/580
* @mebristo made their first contribution in https://github.com/microsoft/InnerEye-DeepLearning/pull/603
* @harshita-s made their first contribution in https://github.com/microsoft/InnerEye-DeepLearning/pull/634
* @fepegar made their first contribution in https://github.com/microsoft/InnerEye-DeepLearning/pull/666
* @peterhessey made their first contribution in https://github.com/microsoft/InnerEye-DeepLearning/pull/725

**Full Changelog**: https://github.com/microsoft/InnerEye-DeepLearning/compare/v0.3...v0.4

<hr /><em>This discussion was created from the release <a href='https://github.com/microsoft/InnerEye-DeepLearning/releases/tag/v0.4'>Environment locking, bug fixes, stability improvements + various other changes.</a>.</em>
